lib/plugins/aws/package/compile/functions/index.js

```diff
@@ -441,98 +442,120 @@ class AwsCompileFunctions {
         fileHash.setEncoding('base64');
         versionHash.setEncoding('base64');
 
-        // Read the file in chunks and add them to the hash (saves memory and performance)
-        versionCompilationPromise = BbPromise.fromCallback(cb => {
-          const readStream = fs.createReadStream(artifactFilePath);
-
-          readStream
-            .on('data', chunk => {
-              fileHash.write(chunk);
-              versionHash.write(chunk);
-            })
-            .on('close', () => {
-              cb();
-            })
-            .on('error', error => {
-              cb(error);
+        // Include function and layer configuration in version id hash (without the Code part)
+        versionCompilationPromise = addFileContentsToHashes(artifactFilePath, [
+          fileHash,
+          versionHash,
+        ])
+          .then(() => {
+            // Include all referenced layer code in the version id hash
+            const layerConfigurations = extractLayerConfigurationsFromFunction(
+              functionResource.Properties,
+              cfTemplate
+            );
+
+            const layerZipFilePaths = [];
+            layerConfigurations.forEach(layer => {
+              const layerZipFileName = layer.properties.Content.S3Key.split('/').pop();
+              const layerZipFilePath = path.join(this.packagePath, layerZipFileName);
+              layerZipFilePaths.push(layerZipFilePath);
             });
-        }).then(() => {
-          // Include function configuration in version id hash (without the Code part)
-          const properties = _.omit(_.get(functionResource, 'Properties', {}), 'Code');
-          _.forOwn(properties, value => {
-            if (_.isObject(value)) {
-              versionHash.write(JSON.stringify(value));
-            } else {
-              versionHash.write(value != null ? String(value) : '');
+
+            return BbPromise.map(layerZipFilePaths, zipFileName => {
```

Comment:
Let's not use Bluebird` functions which do not have native counterparts.

We should use `BbPromise.all` instead

Reply:
This has been updated now, but I realized that both approaches in their current state will run asynchronously on the same hash object. I'm not sure if there could be a race condition or something if `hash.write(chunk)` calls on the same object collide.

A potential fix would be to rewrite `addFileContentsToHash` to return a new string or hash object representing the file. Once all promises finish, the array of strings/hashes could then be combined and added to the versionId hash. What do you think?

Reply:
> This has been updated now, but I realized that both approaches in their current state will run asynchronously on the same hash object.

That's true, it won't be deterministic then. We need to run it sequentially, I believe you can achieve it as:

```javascript
return layerZipFilePaths.reduce(
  (previousHashCalculation, zipFileName) => previousHashCalculation.then(() => addFileContentsToHashes(zipFileName, [versionHash]),
  BbPromise.resolve()
)
```

Note that also `layerZipFilePaths` order needs to also be deterministic

Reply:
Thanks, this should be done now.